packages/core/http/core-http-router-server-internal/src/versioned_router/inject_response_headers.ts

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { IKibanaResponse } from '@kbn/core-http-server';
+import { KibanaResponse } from '../response';
+
+/**
+ * We "copy" to inject our headers without mutating the original response.
+ * @internal
+ */
+export function injectResponseHeaders(headers: object, response: IKibanaResponse): IKibanaResponse {
+  return new KibanaResponse(response.status, response.payload, {
+    ...response.options,
+    headers: {
+      ...response.options?.headers,
```

Comment:
I thought it was a POJSO, but then I ran into this: https://github.com/elastic/kibana/blob/b75d89a2d86c976d60144f8c57b1d415b2f35ac3/packages/core/http/core-http-router-server-internal/src/response_adapter.ts#L61

We could just mutate the value in the same way we do with the request to avoid constructing again.